scripts/broadcast-nmdp-discover.nse

```diff
@@ -0,0 +1,231 @@
+--       socket timeout (default: 5s)
+---
+
+author = "Brendan Coles"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery", "broadcast", "safe"}
+
+prerule = function() return ( nmap.address_family() == "inet") end
+
+local arg_address = stdnse.get_script_args(SCRIPT_NAME .. ".address")
+local arg_timeout = stdnse.parse_timespec(stdnse.get_script_args(SCRIPT_NAME .. ".timeout"))
+
+-- Listens for NMDP response messages.
+--@param interface Network interface to listen on.
+--@param timeout Time to listen for a response.
+--@param responses table to insert response data into.
+local nmdpListen = function(interface, timeout, responses)
+  local condvar = nmap.condvar(responses)
+  local start = nmap.clock_ms()
+  local listener = nmap.new_socket()
+  local status, l3data
+  local filter = 'udp src port 5678 and udp dst port 5678 and src host not ' .. interface.address
+  listener:set_timeout(500)
+  listener:pcap_open(interface.device, 1024, true, filter)
+
+  while (nmap.clock_ms() - start) < timeout do
+    status, _, _, l3data = listener:pcap_receive()
+
+    if not status then
+      goto continue
+    end
+
+    local p = packet.Packet:new(l3data, #l3data)
+    local data = l3data:sub(p.udp_offset + 9)
+
+    if #data < 4 then
+      goto continue
+    end
+
+    stdnse.print_debug(1, "Received NMDP response from %s (%s bytes)", p.ip_src, string.len(data))
+
+    local tlv_type, tlv_len, tlv_value, pos
+    pos = 1
+    --local header = data:sub(pos, pos + 1)
+    pos = pos + 2
+    --local seqno = data:sub(pos, pos + 1)
+    pos = pos + 2
+
+    local mac_address, identity, version, platform, uptime, software_id, board, unpacking, interface
+    while (pos < #data) do
+      -- TLV Type - Unsigned integer (2 bytes)
+      tlv_type = string.unpack(">I2", data:sub(pos, pos + 1))
+      pos = pos + 2
+
+      -- TLV Length - Unsigned integer (2 bytes)
+      tlv_len  = string.unpack(">I2", data:sub(pos, pos + 1))
+      pos = pos + 2
+
+      -- TLV Value
+      tlv_value = data:sub(pos, pos + tlv_len - 1)
+      pos = pos + tlv_len
+
+      --stdnse.print_debug(2, "TLV Type: %s", tlv_type)
+      --stdnse.print_debug(2, "TLV Length: %s", tlv_len)
+      --stdnse.print_debug(2, "TLV Value: %s", stdnse.tohex(tlv_value))
+
+      -- MAC address
+      if tlv_type == 0x01 then
+        mac_address = stdnse.format_mac(tlv_value)
+
+      -- Identity
+      elseif tlv_type == 0x05 then
+        identity = tlv_value
+
+      -- Version
+      elseif tlv_type == 0x07 then
+        version = tlv_value
+
+      -- Platform
+      elseif tlv_type == 0x08 then
+        platform = tlv_value
+
+      -- Uptime - unsigned integer (big endian)
+      elseif tlv_type == 0x0a then
+        uptime_num = string.unpack("<I", tlv_value)
+        local h = math.floor(uptime_num / 3600)
+        local m = math.floor((uptime_num - (h * 3600)) / 60)
+        uptime = h .. "h" .. m .. "m"
```

Comment:
Instead of this math,  you can just pass a number of seconds to [`datetime.format_time()`](https://nmap.org/nsedoc/lib/datetime.html#format_time).

Reply:
Keeping with tradition of existing broadcast scripts, the script sends two broadcast packets "just in case".

Using `datetime.format_time()` introduces `seconds` into the `uptime` string, which causes duplicates in the results table. Comparatively, using only `hour` and `minute` decreases the chance of duplicates. Granted, this approach does run the risk of duplicates should you happen to run the script close to a minute epoch.